### PR TITLE
Adding Catalyst 2.0 smoke test

### PR DESCRIPTION
**Pull request by ayenpure**
-- Adding a smoke test for Catalyst
-- Updating additional versions for Catalyst

### CODE REVIEW

**Comment by ayenpure** [comment]
@kwryankrattiger @mathstuf Can you please review these changes?

**Comment by mathstuf** [comment]
Seems reasonable. See also E4S-Project/testsuite#65

**Review by mathstuf — APPROVED** [review]

File: var/spack/repos/builtin/packages/libcatalyst/package.py [hunk 1]
```diff
@@ -14,10 +18,16 @@ class Libcatalyst(CMakePackage):
...
+    version("master", branch="master")
+    version(
+        "2.0.0-rc4",
+        sha256="cb491e4ccd344156cc2494f65b9f38885598c16d12e1016c36e2ee0bc3640863",
+        preferred=True,
+    )
     version("2.0.0-rc3", sha256="8862bd0a4d0be2176b4272f9affda1ea4e5092087acbb99a2fe2621c33834e05")
-
+    version("2.0.0-rc2", sha256="661ae7c630ff6e382dba45b90bc27e7440e0ed27a2c9dec23c053cec4c7d9507")
+    version("2.0.0-rc1", sha256="451952ca196b904347e9e03a66dedc335d4420d71f3e80d6695aea22fa12e824")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/libcatalyst/package.py]: Why are we adding older release candidates? Aren't these versions supposed to be "ephemeral" and disappear once v2.0.0 is released?

**Review by alalazo — CHANGES_REQUESTED** [review]
There seem to be some legit failures in pipelines

File: var/spack/repos/builtin/packages/libcatalyst/package.py [hunk 2]
```diff
@@ -36,3 +46,28 @@ def cmake_args(self):
         ]
 
         return args
+
+    def smoke_test(self):
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/libcatalyst/package.py]: Is your intention to eventually re-use this method as a stand-alone test as well as a build-time test?

See the docs at https://spack.readthedocs.io/en/latest/packaging_guide.html#checking-an-installation.

**Review by tldahlgren — COMMENTED** [review]
Confirmed the branch and three version sha256 but cannot install the package with tests enabled.

File: var/spack/repos/builtin/packages/libcatalyst/package.py (hunk 2 above)
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/libcatalyst/package.py]: FWIW. I am unable to build the package using `spack -v install --test=root`:

```
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: $SPACK_ROOT/lib/spack/env/gcc/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at CMakeLists.txt:13 (find_package):
  By not providing "Findcatalyst.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "catalyst",
  but CMake did not find one.

  Could not find a package configuration file provided by "catalyst"
  (requested version 2.0) with any of the following names:

    catalystConfig.cmake
    catalyst-config.cmake

  Add the installation prefix of "catalyst" to CMAKE_PREFIX_PATH or set
  "catalyst_DIR" to a directory containing one of the above files.  If
  "catalyst" provides a separate development package or SDK, be sure it has
  been installed.


-- Configuring incomplete, errors occurred!
```

**Review by tldahlgren — CHANGES_REQUESTED** [review]
Escalating the test-related build failure to indicate the problem should be addressed.

File: var/spack/repos/builtin/packages/libcatalyst/package.py (hunk 1 above)
Comment (kwryankrattiger) [reply var/spack/repos/builtin/packages/libcatalyst/package.py]: I think you are right, we should probably only have the latest available here.

On that note, the datestamp release I don't think is usable anymore, it is more of a Beta/Alpha release, so we could drop that as well.

File: var/spack/repos/builtin/packages/libcatalyst/package.py (hunk 2 above)
Comment (kwryankrattiger) [reply var/spack/repos/builtin/packages/libcatalyst/package.py]: Ideally this should be usable as a stand-alone test, but requires installing the examples which is currently not implemented in the RC.

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/libcatalyst/package.py]: One option is to save the examples off after `install` (see https://spack.readthedocs.io/en/latest/packaging_guide.html#saving-build-time-files) then build and run the tests when those files are copied to the test stage directory.

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/libcatalyst/package.py]: > Ideally this should be usable as a stand-alone test, but requires installing the examples which is currently not implemented in the RC.

In that case, why have a separate `smoke_test` method?

File: var/spack/repos/builtin/packages/libcatalyst/package.py [hunk 3]
```diff
@@ -36,3 +46,28 @@ def cmake_args(self):
...
+            cmake(*(["--build", "."]))
+            tty.info("Running Catalyst test")
+
+            res = subprocess.run(["adaptor0/adaptor0_test", "catalyst"])
+            assert res.returncode == 0
+
+    @on_package_attributes(run_tests=True)
+    @run_after("install")
+    def build_test(self):
+        self.smoke_test()
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/libcatalyst/package.py]: Under the circumstances, you can move the contents of `smoke_test()` into this method.

File: var/spack/repos/builtin/packages/libcatalyst/package.py [hunk 4]
```diff
@@ -36,3 +46,28 @@ def cmake_args(self):
...
         return args
+
+    def smoke_test(self):
+        spec = self.spec
+        testdir = "smoke_test_build"
+
+        with working_dir(testdir, create=True):
+            cmake = Executable(spec["cmake"].prefix.bin.cmake)
+            cmakeExampleDir = join_path(self.stage.source_path, "examples")
+            catalystPath = join_path(spec["libcatalyst"].prefix.lib.cmake, "catalyst-2.0")
```
Comment (mathstuf) [inline var/spack/repos/builtin/packages/libcatalyst/package.py]: The examples are failing to use this directory, just pass the prefix through `CMAKE_PREFIX_PATH` and let CMake search inside there.

I think if there's an external conduit, that prefix should be passed as well.

File: var/spack/repos/builtin/packages/libcatalyst/package.py (hunk 3 above)
Comment (ayenpure) [reply var/spack/repos/builtin/packages/libcatalyst/package.py]: Done! :)

File: var/spack/repos/builtin/packages/libcatalyst/package.py (hunk 1 above)
Comment (ayenpure) [reply var/spack/repos/builtin/packages/libcatalyst/package.py]: I've removed all the previous versions.

**Comment by kwryankrattiger** [comment]
@spackbot run pipeline

**Review by tldahlgren — APPROVED** [review]
This works.  The build log shows:

```
...
-- Build files have been written to: /tmp/dahlgren/spack-stage/spack-stage-libcatalyst-2.0.0-rc4-vofrtfkts3c2jqtpfygs2zd3e2mifspj/spack-src/smoke_test_build
==> [2023-09-27-17:26:27.891354] $SPACK_ROOT/spack/opt/spack/linux-rhel8-x86_64_v3/gcc-12.1.1/cmake-3.26.3-xhxkjrk6lpvyajpdocyw7q5y2yfrfnre/bin/cmake' '--build' '.'
[ 14%] Building CXX object about/CMakeFiles/catalyst_example_about.dir/catalyst_example_about.cxx.o
[ 28%] Linking CXX executable catalyst_example_about
[ 28%] Built target catalyst_example_about
[ 42%] Building C object adaptor0/CMakeFiles/catalyst_example_adaptor0.dir/catalyst_impl_example_adaptor0.c.o
[ 57%] Building CXX object adaptor0/CMakeFiles/catalyst_example_adaptor0.dir/catalyst_example_adaptor0.cxx.o
[ 71%] Linking CXX shared module ../catalyst/libcatalyst-example_adaptor0.so
[ 71%] Built target catalyst_example_adaptor0
[ 85%] Building CXX object adaptor0/CMakeFiles/adaptor0_test.dir/adaptor0_test.cxx.o
[100%] Linking CXX executable adaptor0_test
[100%] Built target adaptor0_test
==> [2023-09-27-17:26:30.198580] Running Catalyst test
```

**Comment by tldahlgren** [comment]
One of the CI stacks timed out.

@spackbot run pipeline

**Review by mathstuf — APPROVED** [review]
Pipeline issues seem resolved and the configuration looks cleaner too.

**Review by kwryankrattiger — APPROVED** [review]

**Comment by kwryankrattiger** [comment]
@alalazo this needs another review from you to merge

**Review by alalazo — APPROVED** [review]

**Pull request merged**